pkg/operator/controllers/muo/muo_controller.go

```diff
@@ -0,0 +1,135 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	aroclient "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned"
+	"github.com/Azure/ARO-RP/pkg/operator/controllers"
+	"github.com/Azure/ARO-RP/pkg/util/dynamichelper"
+)
+
+const (
+	enabled  string = "rh.srep.muo.enabled"
+	managed  string = "rh.srep.muo.managed"
+	pullSpec string = "rh.srep.muo.deploy.pullspec"
+)
+
+type Reconciler struct {
+	log           *logrus.Entry
+	arocli        aroclient.Interface
+	kubernetescli kubernetes.Interface
+	dh            dynamichelper.Interface
+	deployer      *deployer
+
+	readinessPollTime time.Duration
+	readinessTimeout  time.Duration
+}
+
+func NewReconciler(log *logrus.Entry, arocli aroclient.Interface, kubernetescli kubernetes.Interface, dh dynamichelper.Interface) *Reconciler {
+	return &Reconciler{
+		log: log,
+
+		arocli:        arocli,
+		kubernetescli: kubernetescli,
+		dh:            dh,
+		deployer:      newDeployer(kubernetescli, dh),
+
+		readinessPollTime: 10 * time.Second,
+		readinessTimeout:  5 * time.Minute,
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	instance, err := r.arocli.AroV1alpha1().Clusters().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !instance.Spec.OperatorFlags.GetSimpleBoolean(enabled) {
+		// controller is disabled
+		return reconcile.Result{}, nil
+	}
+
+	managed := instance.Spec.OperatorFlags.GetWithDefault(managed, "")
+
+	// If enabled and managed=true, install MUO
+	// If enabled and managed=false, remove the MUO deployment
+	// If enabled and managed is missing, do nothing
+	if strings.EqualFold(managed, "true") {
+		imagePullspec, ext := instance.Spec.OperatorFlags[pullSpec]
+		if !ext {
+			return reconcile.Result{}, errors.New("missing pullspec")
+		}
+
+		err = r.deployer.CreateOrUpdate(ctx, imagePullspec)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// Check that MUO has become ready, wait up to readinessTimeout (default 5min)
+		timeoutCtx, cancel := context.WithTimeout(ctx, r.readinessTimeout)
+		defer cancel()
+
+		err := wait.PollImmediateUntil(r.readinessPollTime, func() (bool, error) {
+			return r.deployer.IsReady(ctx)
+		}, timeoutCtx.Done())
+		if err != nil {
+			r.log.Error("Managed Upgrade Operator deployment timed out on Ready")
+			return reconcile.Result{}, err
+		}
+	} else if strings.EqualFold(managed, "false") {
+		errs := r.deployer.Remove(ctx)
+		if len(errs) != 0 {
+			r.log.Error("Error removing MUO:")
+			for _, e := range errs {
+				r.log.Error(e)
+			}
+			// just return one of the errors
+			return reconcile.Result{}, errs[0]
```

Comment:
1. Multiple log calls will mean that it is likely that logs from this conroller will likely interleave with soemthing else. IMO, we need a single log call here for readability (if we want to log explicitly, see below).
2. Controller-runtime already does logging for us. So we will be logging all erorrs here and then controller runtime will log first error again.
3. Handling a slice of errors is tricky and not very useful in this case.

---

Taking above into account, I would like to suggest to change `Remove` signature to return a single error and inside `Remove`:

* Either concatenate all errors into a single error and return it.
* Or use error wrapping (https://pkg.go.dev/errors and https://go.dev/blog/go1.13-errors has a bit of info on wrapping)

And here we will jsut have our normal 

```go
	if err != nil {
		return reconcile.Result{}, err
	}
```

Reply:
Done